search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,447 @@
+        }
+      else if (g_strcmp0 (key, "limit") == 0)
+        {
+          g_value_set_int (append_construction_prop ("limit",
+                                                     G_TYPE_INT,
+                                                     values_array,
+                                                     props_array,
+                                                     &nprops),
+                           g_variant_get_int32 (value));
+        }
+      else if (g_strcmp0 (key, "offset") == 0)
+        {
+          g_value_set_int (append_construction_prop ("offset",
+                                                     G_TYPE_INT,
+                                                     values_array,
+                                                     props_array,
+                                                     &nprops),
+                           g_variant_get_int32 (value));
+        }
+    }
+
+  return EKNC_QUERY_OBJECT (g_object_new_with_properties (EKNC_TYPE_QUERY_OBJECT,
+                                                          nprops,
+                                                          (const char **) props_array->pdata,
+                                                          (const GValue *) values_array->data));
+}
+
+static gboolean
+handle_query (EksArticleMetadata    *skeleton,
+              GDBusMethodInvocation *invocation,
+              GVariant              *query_params,
+              gpointer               user_data)
+{
+  EksArticleMetadataProvider *self = user_data;
+  EkncEngine *engine = eknc_engine_get_default ();
+  g_autoptr(EkncQueryObject) query =
+    create_query_from_dbus_query_parameters(query_params, self->application_id);
+
+  /* Hold the application so that it doesn't go away whilst we're handling
+   * the query */
+  g_application_hold (g_application_get_default ());
+
+  eknc_engine_query (engine,
+                     query,
+                     self->cancellable,
+                     on_received_query_results,
+                     article_metadata_query_state_new (self, invocation));
+  return TRUE;
+}
+
+static gboolean
+handle_shards (EksArticleMetadata    *skeleton,
+               GDBusMethodInvocation *invocation,
+               gpointer               user_data)
+{
+  EksArticleMetadataProvider *self = user_data;
+  EkncEngine *engine = eknc_engine_get_default ();
+  g_autoptr(GError) error = NULL;
+  EkncDomain *domain = eknc_engine_get_domain_for_app (engine,
+                                                       self->application_id,
+                                                       &error);
+  g_auto(GStrv) shards_strv = NULL;
+
+  if (domain == NULL)
+    {
+      g_dbus_method_invocation_take_error (invocation,
+                                           g_steal_pointer (&error));
+
+      /* No need to free_full the out models and shards here,
+       * g_slist_copy_deep is not called if this function returns FALSE. */
```

Comment:
Irrelevant comment

Reply:
Please remove it

Reply:
Oops! I must have forgotten to do this earlier :)